File: webapp/backend/schemas/scenario.py
from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime

class ChatMessage(BaseModel):
    role: str   # "user" | "assistant" | "system"
    content: str

class ScenarioRead(BaseModel):
    id: str
    video_id: str
    source_type: str
    source_content: str | None = None
    chat_messages: str | None = None   # JSON 文字列
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)

class ScenarioOutlineItem(BaseModel):
    """チャットのアウトライン段階での 1 シーン（軽量）。LLM 出力の揺れに強いよう全項目にデフォルトを持たせる。"""
    index: int = Field(default=1)
    title: str = Field(default="")
    summary: str = Field(default="")          # このシーンで扱う内容のあらすじ（1〜2文）
    # 段階 1 で決めるのは「情報の型」（14 択）。具体的な見せ方は
    # シーン内容の生成時（段階 2）に、実際の件数を見てから決める。
    content_type: str = Field(default="")
    # 後方互換。古いフロントや保存済みデータは見せ方の名前を直接送ってくる。
    layout_type: str = Field(default="")

class ScenarioOutline(BaseModel):
    """チャットのアウトライン LLM レスポンスのパース結果"""
    scenes: list[ScenarioOutlineItem]

class FromTextRequest(BaseModel):
    text: str

class ChatRequest(BaseModel):
    message: str

class FinalizeRequest(BaseModel):
    """チャットのアウトライン確定（スケルトンシーン作成）"""
    scenes: list[ScenarioOutlineItem]

class NarrationGenerateRequest(BaseModel):
    pass

class PptxImportResult(BaseModel):
    """PPTX 取り込み直後のレスポンス（ナレーション生成は非同期ジョブで進行する）"""
    scenario: ScenarioRead
    job_id: str
    slide_count: int
    scene_count: int
    visual_count: int
    warnings: list[str] = Field(default_factory=list)

File: webapp/frontend/src/stores/scenario.js
import { defineStore } from 'pinia'
import { scenarioApi } from '@/api/scenario'

export const useScenarioStore = defineStore('scenario', {
  state: () => ({
    scenario: null,          // ScenarioRead | null
    chatMessages: [],        // [{ role, content }]
    lastProposal: null,      // ScenarioProposal | null
    loading: false,
    error: null,
  }),
  actions: {
    async fetchScenario(videoId) {
      this.loading = true
      this.error = null
      try {
        const { data } = await scenarioApi.get(videoId)
        this.scenario = data
        if (data && data.chat_messages) {
          this.chatMessages = JSON.parse(data.chat_messages)
          // 履歴の最後の assistant 返答から proposal の抽出を試みて lastProposal に保持する
          const assistantMsgs = this.chatMessages.filter(m => m.role === 'assistant')
          if (assistantMsgs.length > 0) {
            const lastReply = assistantMsgs[assistantMsgs.length - 1].content
            this.extractLastProposal(lastReply)
          }
        } else {
          this.chatMessages = []
          this.lastProposal = null
        }
      } catch (e) {
        this.error = e.response?.data?.detail || e.message
        throw e;
      } finally {
        this.loading = false
      }
    },

    async fromPptx(videoId, file) {
      this.loading = true
      this.error = null
      try {
        // レスポンスは PptxImportResult { scenario, job_id, slide_count, scene_count, visual_count, warnings }
        // ナレーション生成は非同期ジョブなので、シーン構成自体はこの時点で確定している
        const { data } = await scenarioApi.fromPptx(videoId, file)
        this.scenario = data.scenario
        this.chatMessages = []
        this.lastProposal = null
        return data
      } catch (e) {
        this.error = e.response?.data?.detail || e.message
        throw e
      } finally {
        this.loading = false
      }
    },

    async fromText(videoId, text) {
      this.loading = true
      this.error = null
      try {
        const { data } = await scenarioApi.fromText(videoId, text)
        this.scenario = data
        this.chatMessages = []
        this.lastProposal = null
        return data
      } catch (e) {
        this.error = e.response?.data?.detail || e.message
        throw e
      } finally {
        this.loading = false
      }
    },

    async sendChatMessage(videoId, message) {
      this.loading = true
      this.error = null
      try {
        // 先行してローカル状態に user メッセージを積む (対話レスポンスの向上)
        this.chatMessages.push({ role: 'user', content: message })
        
        const { data } = await scenarioApi.chat(videoId, message)
        
        // ローカル状態を同期 (システムメッセージ等も含めて再同期されるようにする)
        if (this.scenario) {
          const messages = JSON.parse(this.scenario.chat_messages || '[]')
        }
        
        // ストアメッセージを更新
        this.chatMessages.push({ role: 'assistant', content: data.reply })
        this.lastProposal = data.proposal // { scenes: [...] } or null
        
        return data
      } catch (e) {
        // エラー時は追加した user メッセージを削除する等のロールバック
        this.chatMessages.pop()
        this.error = e.response?.data?.detail || e.message
        throw e
      } finally {
        this.loading = false
      }
    },

    async finalizeScenario(videoId, scenes) {
      this.loading = true
      this.error = null
      try {
        const { data } = await scenarioApi.finalize(videoId, scenes)
        this.scenario = data
        this.lastProposal = null
        return data
      } catch (e) {
        this.error = e.response?.data?.detail || e.message
        throw e
      } finally {
        this.loading = false
      }
    },

    extractLastProposal(text) {
      try {
        const match = text.match(/```(?:json)?\s*(\{.*?\})\s*```/s)
        let jsonStr = ''
        if (match) {
          jsonStr = match[1]
        } else {
          const matchRaw = text.match(/(\{.*\})/s)
          if (matchRaw) {
            jsonStr = matchRaw[1]
          }
        }
        if (jsonStr) {
          const parsed = JSON.parse(jsonStr)
          if (parsed && parsed.scenes) {
            this.lastProposal = parsed
          }
        }
      } catch (e) {
        this.lastProposal = null
      }
    }
  }
})
